Initial Implementation

elmigranto opened the pull request:
For #1.

elmigranto committed:
Fixed globs in package.json.

elmigranto committed:
Initial take on `aliases` Couch view.

elmigranto commented:
DB Schema:

* authentication data under `id:<user-id>`;
* a bunch of "aliases" under `alias:<alias-type>:<alias-value>`;
* [`aliases` view](https://github.com/j3k0/ganomede-directory/pull/2/commits/ee951c27d877600a27c45ae7c1e63ab790ec8141#diff-b8aabbee742d7e54ab5af22f330679d6) that we can ask via Couch query param `key` for info on particular user.

We'll have something like this in Couch:
![screenshot 2016-12-26 15 06 19](https://cloud.githubusercontent.com/assets/886388/21480234/e28fead6-cb7c-11e6-9880-a91acc866075.png)

Raw Couch output that will later get formatted by node app:

``` json
GET <DB_URL>/_design/users/_view/aliases?key=%22alice%22&include_docs=true

{
  "total_rows": 4,
  "offset": 0,
  "rows": [
    {
      "id": "alias:email:alice@wonderland.com",
      "key": "alice",
      "value": {
        "_id": "id:alice",
        "public": true,
        "date": "Tue Dec 20 2016 18:06:51 GMT+0500 (+05)",
        "type": "email",
        "value": "alice@wonderland.com"
      },
      "doc": {
        "_id": "id:alice",
        "_rev": "2-ab524315d9bd64be10a02d8f8d80d388",
        "stuff": "things"
      }
    },
    {
      "id": "alias:name:Alice In Wonderland",
      "key": "alice",
      "value": {
        "_id": "id:alice",
        "public": true,
        "date": "Tue Dec 20 2016 18:06:51 GMT+0500 (+05)",
        "type": "name",
        "value": "Alice In Wonderland"
      },
      "doc": {
        "_id": "id:alice",
        "_rev": "2-ab524315d9bd64be10a02d8f8d80d388",
        "stuff": "things"
      }
    }
  ]
}
```

elmigranto commented:
With above creating a user will be a bit awkward — some docs might go missing (no transactional guarantees), some might conflict (like email is okay, but name isn't — email gets inserted), so we'll need to dance around that a little bit.

For basic search we have somewhat direct mapping to underlying Couch doc with known id. For more sophisticated stuff, we can add a view later.

Pagination of users will require additional simple view on ids or similar, shouldn't be a problem.

So overall, looks fine to me, what do you think? Am I missing something?

**upd** I see some issues (like `doc` being on every alias), will fix these, but general direction looks okay.

j3k0 commented:
Seems fine.

For transactional-like, you can do bulk insertions in couchdb, add `"all_or_nothing":true`, it is not 100% like a transaction, as the replication might not be bulkified, but close enough.

elmigranto commented:
From what I've read about `all_or_nothing`, it'll write conflicted versions for already existing IDs, not sure if that's better…

j3k0 commented:
While not 100% necessary, I think you should also add a `reduce()` function which concatenates aliases for each user id, while making sure only the most recent of them is kept in the array.

j3k0 commented:
> From what I've read about all_or_nothing, it'll write conflicted versions for already existing IDs, not sure if that's better…

Okay, it might be better without it then.

elmigranto committed:
First take on user creation, some intial objects.

j3k0 commented on src/users/CreatesUsers.js:
```diff
@@ -0,0 +1,67 @@
+  setImmediate(callback, null, docs);
+};
+
+class CreatesUsers {
```
@elmigranto Out of curiosity, is this some ES6 standard naming convention for classes? 

(we generally use nouns as class names, it's the first time I see verbs conjugated at the 3rd person... No biggy though).

elmigranto replied:
It's nothing special, just a small piece from a thing I am trying out. Don't have strong feelings either way.

j3k0 commented on src/users/CreatesUsers.js:
```diff
@@ -0,0 +1,67 @@
+  }
+}
+
+module.exports = CreatesUsers;
```
Also, in this context, a class with a single static method (or at least "could-be" static, since it's not using any instance field), you might just export the "createUser()" method... no?

elmigranto replied:
So I am trying this "outside-in" design approach with [testdouble.js library](https://github.com/testdouble/testdouble.js) where you split app into a kind of "layered cake": define interactions between objects in higher levels first, while faking out details on lower ones. The lib is good at replacing classes and objects with functions. It's just easier to `td.replace` constructor because it fakes all the methods in one go, and with objects you need to manually replace functions one by one, so I default to classes.

Here's more info if you are interested — http://blog.testdouble.com/posts/2015-09-10-how-i-use-test-doubles.html

j3k0 replied:
Cool thing. I'm pro-outside-in design, consider how a piece of code will be used before considering how it will be implemented generally end up in nicer to use apis.

elmigranto commented:
We may need to alter some endpoints that result in touching multiple Couch docs due to "no transactions" situation.

**User creation**
Multiple calls (create `id:*` doc with id and password, than add aliases, see below) instead of one since we create multiple docs and any of them can 409, but aliases depend on `id:*` doc.

User creation will reply with auth token, which is used to add aliases.

**Adding aliases**
One by one instead of array, or we can 409 per alias instead per request (replying with which aliases succeeded and which failed).

<hr/>
It also seems like simpler endpoints logic-wise with less intersecting concerns. Otherwise we'll need to hack quite a bit on top of Couch, at least for sensitive parts like user creation where docs are related.

What do you think?

j3k0 commented:
That would makes sense: we remove the need for transactions in the directory module so it is all clean inside.

However, the consumer of this API will still need to create a user with a number of aliases when a new user registers... So we are just pushing the resolution of this problem to the layer above (that will be ganomede-users).

So... the "atomic" API you suggest can be added.

Still, in order to make "ganomede-users" just a glue between different modules with the minimal amount of logic, I'd still like to present an API to make this kind of transactional operation.

Concerning failure, what if you implement a kind of "cancellable action" design pattern (it has a real name that I forgot). If 1 fails, the whole request fails and any "executed" action is cancelled.

The idea for "cancellable actions":

```js
class CreateUserId {
  constructor(id, password) { ... }
  execute(cb) { ... } // create document on couchdb
  cancel(cb) { ... } // delete document on couchdb
}

class CreateAlias {
  constructor(userid, key, value) { ... }
  execute(cb) { ... } // create doc on couch
  cancel(cb) { ... } // delete doc on couchdb
}

myRequest = (req,res,next) {
  // prepare list of operations to exectute
  actions = [
    new CreateUserId(...),
    new CreateAlias(...)
  ];
  executeActions(actions, [], next);
};

cancelActions = (actions, cb) {
  if (actions.length == 0) return cb();
  actions[actions.length - 1].cancel(() {
    cancelActions(actions.slice(0, actions.length - 1), 0);
  });
};

executeActions = (actions, executed, cb) {
  // execute operations sequentially until 1 fails or all succeeds
  // if 1 fails, cancel operations in reverse order (starting from failed one)
  if (actions.length == 0) return cb();
  actions[0].execute((err) {
    if (err) return cancelActions(executed, cb);
    executeActions(actions.slice(1), executed.append(actions[0]), cb);
  });
};
```

(made-up code to illustrate this)

There still is a possibility that the server crashes during cancellation... but it's ok, we can live with this.

j3k0 commented on src/users/CreatesUsers.js:
```diff
@@ -0,0 +1,67 @@
+  }
+}
+
+module.exports = CreatesUsers;
```
Cool thing. I'm pro-outside-in design, consider how a piece of code will be used before considering how it will be implemented generally end up in nicer to use apis.

elmigranto commented:
Couple of rough edges I thought of, nothing quite major, but just to keep them in mind:

1. Overwriting/resetting to previous alias. Say someone once was "jack", changed it to "bob", but decided to switch to original "jack" — 409. Will need look up alias ownership to alleviate this.
1. Password changes — simply update document, nothing major.
1. Password resets — more relations and doc updates with one time tokens.

elmigranto committed:
ActionExecutor class for approximating transactions.

elmigranto committed:
Implemented `actions/CreateUserDocument`.

elmigranto committed:
Implemented `actions/CreateAliasDocument`.

elmigranto committed:
CreatesUsers now accesses db via ActionsExecutor.

j3k0 commented:
Seems great so far.

Indeed: ex "jack", now "bob", want to be called "jack" again we should allow him to do so.

elmigranto committed:
Implemented users/LoginsUsers.

elmigranto commented:
> I think you should also add a `reduce()` function which concatenates aliases for each user id

I've been poking this for some time, not sure how to achieve something like that… Firstly, couch doesn't really know who's asking atm — admin who should see all the aliases, or regular user who should see public ones only. Secondly, I'm not quite sure how to properly concatenate stuff (keep getting `reduce_overflow_error` on like 5 documents), so for now I am going with 2 queries that merged and formatted within node into profile object: 1) fetching user document and 2) fetching all user's aliases from view with `key` qs param.

<hr/>

Perhaps you can look through `profiles` view and help me out with issues?

``` js
{ profiles: {
    map: String(function (doc) {
      const parts = doc._id.split(':');
      const type = parts[0];

      switch (type) {
        case 'id': {
          emit(doc.id, {
            id: doc.id,
            hash: doc.hash,
            aliases: []
          });
          break;
        }

        case 'alias': {
          emit(doc.id, {
            id: null,
            hash: null,
            aliases: [{
              public: doc.public,
              date: doc.date,
              type: parts[1],
              value: parts[2]
            }]
          });
          break;
        }

        default:
          break;
      }
    }),

    // should work with `group=true`
    reduce: String(function (key, docs/*, rereduce*/) {
      const profile = {
        id: null,
        hash: null,
        aliases: []
      };

      docs.forEach(function (doc) {
        profile.id = profile.id || doc.id;
        profile.hash = profile.hash || doc.hash;
        profile.aliases = profile.aliases.concat(doc.aliases); // removing this works, but, obviously, useless
      });

      return profile;
    })
  }
}
```

j3k0 commented:
Just a guess so far, but maybe using `id` as a field is looking for trouble here (?)

I'll look into it with more details.

j3k0 commented:
Seems to work fine here (using couchdb 1.6.1)

 * Ran a fresh couchdb container with `docker run --rm -it -p 5984:5984 klaemo/couchdb:1.6.1`
 * Copy-pasted the map and reduce functions above.
 * Inserted a few docs

This is the dump of the DB: `curl "localhost:5984/directory/_all_docs?include_docs=true"`

```json
{
   "total_rows" : 6,
   "rows" : [
      {
         "value" : {
            "rev" : "1-b986d07957b0ad40dde291b677d7bc19"
         },
         "doc" : {
            "views" : {
               "test" : {
                  "map" : "function (doc) {\n      const parts = doc._id.split(':');\n      const type = parts[0];\n\n      switch (type) {\n        case 'id': {\n          emit(doc.id, {\n            id: doc.id,\n            hash: doc.hash,\n            aliases: []\n          });\n          break;\n        }\n\n        case 'alias': {\n          emit(doc.id, {\n            id: null,\n            hash: null,\n            aliases: [{\n              public: doc.public,\n              date: doc.date,\n              type: parts[1],\n              value: parts[2]\n            }]\n          });\n          break;\n        }\n\n        default:\n          break;\n      }\n    }",
                  "reduce" : "function (key, docs/*, rereduce*/) {\n      const profile = {\n        id: null,\n        hash: null,\n        aliases: []\n      };\n\n      docs.forEach(function (doc) {\n        profile.id = profile.id || doc.id;\n        profile.hash = profile.hash || doc.hash;\n        profile.aliases = profile.aliases.concat(doc.aliases); // removing this works, but, obviously, useless\n      });\n\n      return profile;\n    }"
               }
            },
            "language" : "javascript",
            "_id" : "_design/test",
            "_rev" : "1-b986d07957b0ad40dde291b677d7bc19"
         },
         "id" : "_design/test",
         "key" : "_design/test"
      },
      {
         "value" : {
            "rev" : "2-d6609f06790a8193a31a1aa8525b95ba"
         },
         "doc" : {
            "date" : "1/1/1920",
            "_id" : "alias:email:jeko@email.fr",
            "public" : true,
            "id" : "idjeko",
            "_rev" : "2-d6609f06790a8193a31a1aa8525b95ba"
         },
         "id" : "alias:email:jeko@email.fr",
         "key" : "alias:email:jeko@email.fr"
      },
      {
         "value" : {
            "rev" : "1-435f02fdc56bafb807d4f70c92d1f574"
         },
         "doc" : {
            "date" : "1/1/2000",
            "_id" : "alias:username:elm",
            "public" : true,
            "id" : "idelmigranto",
            "_rev" : "1-435f02fdc56bafb807d4f70c92d1f574"
         },
         "id" : "alias:username:elm",
         "key" : "alias:username:elm"
      },
      {
         "value" : {
            "rev" : "1-1f2580777746ea7876556e49dcfbec6a"
         },
         "doc" : {
            "date" : "4/7/1980",
            "_id" : "alias:username:jeko",
            "public" : true,
            "id" : "idjeko",
            "_rev" : "1-1f2580777746ea7876556e49dcfbec6a"
         },
         "id" : "alias:username:jeko",
         "key" : "alias:username:jeko"
      },
      {
         "value" : {
            "rev" : "1-5258ba19d60511889c11a2c25d4283f9"
         },
         "doc" : {
            "hash" : "654231",
            "_id" : "id:idelmigranto",
            "id" : "idelmigranto",
            "_rev" : "1-5258ba19d60511889c11a2c25d4283f9"
         },
         "id" : "id:idelmigranto",
         "key" : "id:idelmigranto"
      },
      {
         "value" : {
            "rev" : "1-e10dab1546c556e38bcb2a6bd6656c0e"
         },
         "doc" : {
            "hash" : "12344",
            "_id" : "id:idjeko",
            "id" : "idjeko",
            "_rev" : "1-e10dab1546c556e38bcb2a6bd6656c0e"
         },
         "id" : "id:idjeko",
         "key" : "id:idjeko"
      }
   ],
   "offset" : 0
}
```

j3k0 commented:
`curl "http://localhost:5984/directory/_design/test/_view/test?group=true"`

```json
{
   "rows" : [
      {
         "value" : {
            "hash" : "654231",
            "aliases" : [
               {
                  "value" : "elm",
                  "date" : "1/1/2000",
                  "public" : true,
                  "type" : "username"
               }
            ],
            "id" : "idelmigranto"
         },
         "key" : "idelmigranto"
      },
      {
         "value" : {
            "hash" : "12344",
            "aliases" : [
               {
                  "value" : "jeko",
                  "date" : "4/7/1980",
                  "public" : true,
                  "type" : "username"
               },
               {
                  "value" : "jeko@email.fr",
                  "date" : "1/1/1920",
                  "public" : true,
                  "type" : "email"
               }
            ],
            "id" : "idjeko"
         },
         "key" : "idjeko"
      }
   ]
}
```

elmigranto commented:
That's wierd, perhaps I should look into my couch settings…

Changed `query_server_config.reduce_limit` to `false` on my machine, seems to work fine, we'll read up on that setting later. What does docker image have there (`<DB_URL>/_utils/config.html`)?

j3k0 commented:
`curl -s "http://localhost:5984/_config"|json_pp`

```json
{
   "query_server_config" : {
      "os_process_limit" : "25",
      "reduce_limit" : "true"
   },
   "compaction_daemon" : {
      "check_interval" : "300",
      "min_file_size" : "131072"
   },
   "view_compaction" : {
      "keyvalue_buffer_size" : "2097152"
   },
   "daemons" : {
      "auth_cache" : "{couch_auth_cache, start_link, []}",
      "uuids" : "{couch_uuids, start, []}",
      "stats_aggregator" : "{couch_stats_aggregator, start, []}",
      "replicator_manager" : "{couch_replicator_manager, start_link, []}",
      "compaction_daemon" : "{couch_compaction_daemon, start_link, []}",
      "index_server" : "{couch_index_server, start_link, []}",
      "query_servers" : "{couch_query_servers, start_link, []}",
      "httpd" : "{couch_httpd, start_link, []}",
      "stats_collector" : "{couch_stats_collector, start, []}",
      "vhosts" : "{couch_httpd_vhost, start_link, []}",
      "external_manager" : "{couch_external_manager, start_link, []}",
      "os_daemons" : "{couch_os_daemons, start_link, []}"
   },
   "query_servers" : {
      "javascript" : "/usr/local/bin/couchjs /usr/local/share/couchdb/server/main.js",
      "coffeescript" : "/usr/local/bin/couchjs /usr/local/share/couchdb/server/main-coffee.js"
   },
   "couch_httpd_auth" : {
      "authentication_db" : "_users",
      "require_valid_user" : "false",
      "authentication_redirect" : "/_utils/session.html",
      "timeout" : "600",
      "allow_persistent_cookies" : "false",
      "iterations" : "10",
      "auth_cache_size" : "50"
   },
   "httpd_db_handlers" : {
      "_design" : "{couch_httpd_db, handle_design_req}",
      "_compact" : "{couch_httpd_db, handle_compact_req}",
      "_view_cleanup" : "{couch_mrview_http, handle_cleanup_req}",
      "_temp_view" : "{couch_mrview_http, handle_temp_view_req}",
      "_changes" : "{couch_httpd_db, handle_changes_req}",
      "_all_docs" : "{couch_mrview_http, handle_all_docs_req}"
   },
   "couch_httpd_oauth" : {
      "use_users_db" : "false"
   },
   "log" : {
      "include_sasl" : "true",
      "level" : "info",
      "file" : "/dev/null"
   },
   "database_compaction" : {
      "checkpoint_after" : "5242880",
      "doc_buffer_size" : "524288"
   },
   "httpd_global_handlers" : {
      "_replicate" : "{couch_replicator_httpd, handle_req}",
      "/" : "{couch_httpd_misc_handlers, handle_welcome_req, <<\"Welcome\">>}",
      "_config" : "{couch_httpd_misc_handlers, handle_config_req}",
      "_utils" : "{couch_httpd_misc_handlers, handle_utils_dir_req, \"/usr/local/share/couchdb/www\"}",
      "_active_tasks" : "{couch_httpd_misc_handlers, handle_task_status_req}",
      "_session" : "{couch_httpd_auth, handle_session_req}",
      "_db_updates" : "{couch_dbupdates_httpd, handle_req}",
      "_log" : "{couch_httpd_misc_handlers, handle_log_req}",
      "favicon.ico" : "{couch_httpd_misc_handlers, handle_favicon_req, \"/usr/local/share/couchdb/www\"}",
      "_all_dbs" : "{couch_httpd_misc_handlers, handle_all_dbs_req}",
      "_restart" : "{couch_httpd_misc_handlers, handle_restart_req}",
      "_oauth" : "{couch_httpd_oauth, handle_oauth_req}",
      "_plugins" : "{couch_plugins_httpd, handle_req}",
      "_uuids" : "{couch_httpd_misc_handlers, handle_uuids_req}",
      "_stats" : "{couch_httpd_stats_handlers, handle_stats_req}"
   },
   "couchdb" : {
      "os_process_timeout" : "5000",
      "attachment_stream_buffer_size" : "4096",
      "max_document_size" : "4294967296",
      "uuid" : "604481753f82a2ca30c02d946948ec3a",
      "uri_file" : "/usr/local/var/run/couchdb/couch.uri",
      "max_dbs_open" : "100",
      "view_index_dir" : "/usr/local/var/lib/couchdb",
      "util_driver_dir" : "/usr/local/lib/couchdb/erlang/lib/couch-1.6.1/priv/lib",
      "database_dir" : "/usr/local/var/lib/couchdb",
      "plugin_dir" : "/usr/local/lib/couchdb/plugins",
      "delayed_commits" : "true",
      "file_compression" : "snappy"
   },
   "uuids" : {
      "algorithm" : "sequential",
      "max_count" : "1000"
   },
   "attachments" : {
      "compressible_types" : "text/*, application/javascript, application/json, application/xml",
      "compression_level" : "8"
   },
   "ssl" : {
      "ssl_certificate_max_depth" : "1",
      "verify_ssl_certificates" : "false",
      "port" : "6984"
   },
   "stats" : {
      "samples" : "[0, 60, 300, 900]",
      "rate" : "1000"
   },
   "httpd" : {
      "vhost_global_handlers" : "_utils, _uuids, _session, _oauth, _users",
      "secure_rewrites" : "true",
      "authentication_handlers" : "{couch_httpd_oauth, oauth_authentication_handler}, {couch_httpd_auth, cookie_authentication_handler}, {couch_httpd_auth, default_authentication_handler}",
      "port" : "5984",
      "default_handler" : "{couch_httpd_db, handle_request}",
      "log_max_chunk_size" : "1000000",
      "enable_cors" : "false",
      "allow_jsonp" : "false",
      "bind_address" : "0.0.0.0",
      "socket_options" : "[{recbuf, 262144}, {sndbuf, 262144}]"
   },
   "cors" : {
      "credentials" : "false"
   },
   "replicator" : {
      "ssl_certificate_max_depth" : "3",
      "db" : "_replicator",
      "connection_timeout" : "30000",
      "worker_processes" : "4",
      "retries_per_request" : "10",
      "max_replication_retry_count" : "10",
      "worker_batch_size" : "500",
      "socket_options" : "[{keepalive, true}, {nodelay, false}]",
      "http_connections" : "20",
      "verify_ssl_certificates" : "false"
   },
   "httpd_design_handlers" : {
      "_update" : "{couch_mrview_show, handle_doc_update_req}",
      "_show" : "{couch_mrview_show, handle_doc_show_req}",
      "_info" : "{couch_mrview_http, handle_info_req}",
      "_list" : "{couch_mrview_show, handle_view_list_req}",
      "_view" : "{couch_mrview_http, handle_view_req}",
      "_compact" : "{couch_mrview_http, handle_compact_req}",
      "_rewrite" : "{couch_httpd_rewrite, handle_rewrite_req}"
   },
   "vendor" : {
      "version" : "1.6.1",
      "name" : "The Apache Software Foundation"
   }
}
```

elmigranto committed:
Couch user profiles: aliases.view -> users.design (with `list` function instead of `reduce` one)

elmigranto commented:
Configs seems quite similar (actually, identical except for stuff like paths). And I think that the problem is that array grows compared to originally emitted documents (unlike, say, reduce functions like `_sum`). Maybe we can solve this with disabling `reduce_limit`, but I'm not sure this is such a great idea, unless there is an option to do that per-design doc.

Common solution seems to be to emit array elements, use `startkey` / `endkey`, and to not `reduce`. But since our array is embedded, not for us. The other option is to use `list` functions, [here's my take on that](https://github.com/j3k0/ganomede-directory/pull/2/commits/5223962bf55373b0c1cc0fe8b379300990c81d22), but won't allow us to enumerate users without making a bunch of requests and craeting additional view to list `userId`s. May or may not work, depending on why we need an endpoint that fetches all the profiles with limit/offset. Let me know and I'll look into this more, or perhaps we'll just add someting like `_views/userIds`.

elmigranto committed:
Added `FindsProfiles` for user lookups based on various identifiers.

elmigranto committed:
users: Profile and BuildsProfiles.

elmigranto committed:
Implemented ChangesPasswords.

elmigranto committed:
Implemented AddsAliases.

elmigranto commented:
I just realized couch doesn't delete documents, but rather soft/hard marks them as `_deleted`, so action rollbacks will leave garbage behind (and using `_purge` sounds like trouble). Probably not an issue, but worth mentioning.

elmigranto committed:
Implemented Db methods.

j3k0 commented:
> action rollbacks will leave garbage behind

Not an issue. However I think we should make sure rollbacks are used as rarely as possible. Maybe a _check_ step ahead of executing the actions would be a good idea?

```js
CreateUserIdAction = {
  execute: (cb) => // create some doc
  rollback: (cb) => // delete the doc
  check: (cb) => // fail if doc already exists
}
```

j3k0 commented:
About the above, let's first finish this and proceed with the rest of the work that relies on this module being done.

What is left to do in this module?

elmigranto commented:
1. ~~Implement going through users list ([can you elaborate on how we plan to use it](https://github.com/j3k0/ganomede-directory/pull/2#issuecomment-274290512)? perhaps we only need user IDs in the list and can fetch profiles one by one when needed?)~~;
1. ~~implement action checks~~;
1. ~~implement access rights and formatting on profiles (`public` thingy, sorting and formatting output)~~;
1. tie everything up to server endpoints.

Everything except *1* is pretty straightforward, should finish it today.

j3k0 commented:
> Implement going through users list

Can be dropped. _I though it'd be necessary for a module that feed MailChimp from the content of directory, as it turns out, MailChimp already has user data: only changes to existing data will have to be taken into account._

> action checks

Ok, I think this is definitely important, but we'll get back to adding this extra polish when all is in place. I'll make a separate issue for this.

> should finish it today.

Great!

elmigranto committed:
Implemented running `check()`s before `execute()`ing actions.

elmigranto committed:
Remove paging through profiles from README.

elmigranto committed:
Implemented Profile object.

elmigranto commented:
Doing endpoints, user creation feels a bit wierd in that it both, creates user and logs him in, so it might be kinda awkward for client to distinguish between user creation error and login error. Maybe split it into two requests clientside — first it creates user and reports any 409, then logs him in (it knows userId and password already).

elmigranto commented:
Also, adding aliases — do we really need the ability to add them in multiples? If so, I'll modify that to run via `ActionExecutor` as well with checks/rollbacks.

elmigranto committed:
Implemented most of endpoints (no API_SECRET yet, though).

elmigranto committed:
Add API_SECRET to some calls.

elmigranto commented:
So I don't think we'll need to add any auth faking (like making token `<API_SECRET>:userId`, simply add `secret` field to query string or json body. Though, maybe we want to do this anyway, since that's what in other modules, can't remember exactly.

elmigranto commented:
All right, almost done, let's decide on above comments and finish this up!

j3k0 commented:
> split user creation and login into 2 requests

OK

> adding aliases in multiples

Might not be 100% necessary. We can have a atomic request for now.

But if I'm right I think we will need the client to create multiple aliases at a time, while making sure they either all get created or not created. Since "directory" already has logic in place to achieve that, adding a "batch" alias creation request will make sense.

> auth faking

If I'm right it's built-in in ganomede-helpers. It already proved useful to keep "auth faking" a general mechanism that works on all modules the same way.

elmigranto committed:
Removed login step from user creation.

elmigranto committed:
Added ability to insert multiple aliases:
  - users/AddsAliases now accepts array of aliases;
  - actions/CreateAliasDocument can overwrite previous aliases by same owner.

elmigranto committed:
Fixing CI: node version to 6.

elmigranto committed:
Fixing CI: add linting step.

elmigranto committed:
Hook everything up to endpoints and index.js.

elmigranto committed:
Fixed some typos.

elmigranto committed:
User ID is `id`, and not `user_id` any longer.

elmigranto commented:
All right, seems reasonably ready! I dumped some things into issues so we don't forget what's "missing", most of them are not serious, except `DbInitializer` in #6 and auth faking #7, which is a bit deeper that I originally thought.

I'll look into those two during rest of the week, but all the functionality is ready and maybe working on `directory` dependents will help us figure out remaining stuff.

To work around DB thing, create `_design/users` couch doc (from [`./src/db/users.design.js`](https://github.com/j3k0/ganomede-directory/blob/91a7b25c05f44278c3cf5e52ee642e80226b86c9/src/db/users.design.js)):

``` json
{
  "_id": "_design/users",
  "views": {
    "rawProfiles": {
      "map": "function (doc) {\n        const parts = doc._id.split(':');\n        const type = parts[0];\n\n        switch (type) {\n          case 'id': {\n            emit(doc.id, {\n              id: doc.id,\n              hash: doc.hash,\n              aliases: []\n            });\n            break;\n          }\n\n          case 'alias': {\n            emit(doc.id, {\n              id: null,\n              hash: null,\n              aliases: [{\n                public: doc.public,\n                date: doc.date,\n                type: parts[1],\n                value: parts[2]\n              }]\n            });\n            break;\n          }\n\n          default:\n            break;\n        }\n      }"
    }
  },
  "lists": {
    "profiles": "function (head, req) {\n      // TODO\n      // Check that `key` query string param is present (`req.query.key`),\n      // meaning we want profile of that userId; throw Error otherwise.\n      // (because this does not make sense without it and will go over all docs).\n      const profile = {\n        id: null,\n        hash: null,\n        aliases: []\n      };\n\n      // `row` is w/ever `emit()` call get translated to by Couch.\n      // `doc` is 2-nd argument to `emit()`.\n      var row; // eslint-disable-line no-var, no-unused-vars\n      var doc; // eslint-disable-line no-var\n\n      while (row = getRow()) {\n        doc = row.value;\n        profile.id = profile.id || doc.id;\n        profile.hash = profile.hash || doc.hash;\n        profile.aliases = profile.aliases.concat(doc.aliases);\n      }\n\n      send(JSON.stringify(profile));\n    }"
  }
}
```

j3k0 commented:
Nice, good job! If auth faking is complicated, we'll push that to later. DbInitializer, however, sounds necessary.

elmigranto committed:
Added DbInitializer.

elmigranto committed:
Split master/worker files and make master init db before forking.

elmigranto committed:
Configure design name via COUCH_DESIGN env var.

elmigranto committed:
Configure running DB sync as script via COUCH_SYNC env var.

elmigranto commented:
Okay, initializer is ready per #6, let's see about fake auth #7 when we start using this module from other ones, maybe we won't even need it right away.

j3k0 committed:
Add docker support

j3k0 commented:
@elmigranto I added `Dockerfile` and `docker-compose.yml`. The app won't start.

I didn't look much into it, any idea?

```
app_1        | /home/app/code/src/db/DbInitializer.js:10
app_1        |     const designsStatus = Object.entries(designs).map(([name, status]) => {
app_1        |                                  ^
app_1        |
app_1        | TypeError: Object.entries is not a function
app_1        |     at OutOfSyncError (/home/app/code/src/db/DbInitializer.js:10:34)
app_1        |     at check (/home/app/code/src/db/DbInitializer.js:127:25)
app_1        |     at /home/app/code/node_modules/async/dist/async.js:356:16
app_1        |     at nextTask (/home/app/code/node_modules/async/dist/async.js:5057:29)
app_1        |     at Immediate.<anonymous> (/home/app/code/node_modules/async/dist/async.js:5064:13)
app_1        |     at apply (/home/app/code/node_modules/async/dist/async.js:21:25)
app_1        |     at Immediate.<anonymous> (/home/app/code/node_modules/async/dist/async.js:56:12)
app_1        |     at Immediate.<anonymous> (/home/app/code/node_modules/async/dist/async.js:840:16)
app_1        |     at Immediate.immediate._onImmediate (timers.js:590:18)
app_1        |     at tryOnImmediate (timers.js:543:15)
```

elmigranto commented:
Yeah, `Object.entries` is behind harmony flag on node 6, gotta use lodash.

elmigranto committed:
Replaced `Object.entries` / `Object.values` with `lodash` alternatives.

j3k0 commented on src/directory.router.js:
```diff
@@ -0,0 +1,143 @@
+const badUserId = next => sendHttpError(next, new restify.BadRequestError('Invalid User ID'));
+const badPassword = next => sendHttpError(next, new restify.BadRequestError('Password must be at least 8 characters long'));
+const badAliases = next => sendHttpError(next, new restify.BadRequestError('Some of the aliases are invalid'));
+
```
Can you make those 3 `bad*` calls emit custom `RestError`, with status `400 (BadRequest)` but different `restCode` (like `"BadUserId"`, `"BadPassword"`, `"BadAliases"`). From [restify's error handling doc](http://restify.com/#error-handling):

```js
new restify.RestError({
    restCode: 'MyError',
    statusCode: 418,
    message: message
});
`

j3k0 replied:
... and if there are other calls that can fail for multiple reasons like those, same idea: would be nice to drop a line one the README.md concerning those calls, about which `restCode` could be expected.

elmigranto replied:
Okay, please check out #9 for changes / discussion on this.

elmigranto committed:
Closes #10.

elmigranto committed:
Changed `sendError` to include error codes.

elmigranto committed:
Changed our `./errors` to have “code” property when converting to restify-supported format:
  - new BaseError#toRestError() method…
  - …which is used by `sendHttpError()`;
  - new class `RequestValidationError` that allows to specify custom error codes for things different things like `BadUserId` and `BadPassword` (basically `restify.BadRequestError` wrapper thing).

elmigranto committed:
Specifying both, new password and new aliases, when editing user is error.

elmigranto committed:
Mention most (hopefully, all) error codes in README specs for endpoints.

elmigranto committed:
errors: change `statusCode` to be a prop rather than getter.
